akka-actor/src/main/scala/akka/util/FrequencySketch.scala

```diff
@@ -0,0 +1,398 @@
+    MurmurHash3.finalizeHash(MurmurHash3.mixLast(hash, hash), 2)
+
+  private def getCounter(row: Int, hash: Int): Int = {
+    val column = (hash & columnMask) >>> slotShift
+    val slot = (hash & slotMask) << counterShift
+    ((matrix(row)(column) >>> slot) & counterMask).toInt
+  }
+
+  private def incrementCounter(row: Int, hash: Int): Boolean = {
+    val column = (hash & columnMask) >>> slotShift
+    val slot = (hash & slotMask) << counterShift
+    val mask = counterMask << slot
+    if ((matrix(row)(column) & mask) != mask) {
+      matrix(row)(column) += (1L << slot)
+      rowSizes(row) += 1
+      true
+    } else false
+  }
+
+  /**
+   * The TinyLFU reset operation (periodically halving all counters).
+   * Adjusts for truncation from integer division (bit shift for efficiency)
+   * by adjusting for the number of odd counters per row (each off by 0.5).
+   */
+  private def reset(): Unit = {
+    var row = 0
+    while (row < depth) {
+      var column = 0
+      var odd = 0
+      while (column < rowWidth) {
+        odd += java.lang.Long.bitCount(matrix(row)(column) & oddMask)
+        matrix(row)(column) = (matrix(row)(column) >>> 1) & resetMask
+        column += 1
+      }
+      rowSizes(row) = (rowSizes(row) - odd) >>> 1
+      row += 1
+    }
+    updatedSize = rowSizes.max
+  }
+
+  def toDebugString: String = FrequencySketchUtil.debugString(matrix, rowWidth, slots, counterWidth, counterMask)
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] object FastFrequencySketch {
+
+  /**
+   * Create a new FastFrequencySketch based on the cache capacity (which will be increased to the nearest power of two).
+   *
+   * @param capacity the cache capacity (maximum items that will be cached)
+   * @param widthMultiplier a multiplier for the width of the sketch
+   * @param resetMultiplier the multiplier on the capacity for the reset size
+   * @return a configured FastFrequencySketch
+   */
+  def apply[A](capacity: Int, widthMultiplier: Int = 4, resetMultiplier: Double = 10): FastFrequencySketch[A] = {
+    val width = widthMultiplier * FrequencySketch.Bits.ceilingPowerOfTwo(capacity)
+    val resetSize = (resetMultiplier * capacity).toInt
+    new FastFrequencySketch(width, resetSize)
+  }
+}
+
+/**
+ * INTERNAL API
+ *
+ * A faster implementation of the frequency sketch (around twice as fast).
+
+ * This frequency sketch uses a fixed depth (number of hash functions) of 4 and a counter size of 4 bits (0-15),
+ * so that constants can be used for improved efficiency. It also uses its own rehashing of item hash codes.
+ *
+ * The implementation is inspired by the approach used in the Caffeine caching library:
+ * https://github.com/ben-manes/caffeine
+ *
+ * @param width width of the count-min sketch (number of counters)
+ * @param resetSize the size (number of counter increments) to apply the reset operation
+ */
+@InternalApi
+private[akka] final class FastFrequencySketch[A](width: Int, resetSize: Int) {
```

Comment:
do you think the CountMinSketch in Artery can be replaced by this?

Reply:
Could be possible. The use case in Artery is maintaining a list of heavy hitters for compression. The approach depends on the count-min sketch having total counts, so it can easily and efficiently see if something has become a heavy hitter. This sketch has a reset operation which halves all the counters, allowing smaller counters and better space efficiency, while still being able to estimate which of two items is heavier, but doesn't retain this total count. We would need to look at using this frequency sketch in a different way — one approach could be to have a frequency list of the heavy hitters and use the frequency sketch as an admission filter (in the same way it's used in the TinyLFU caching policy) to determine whether to swap out the least frequent position. And where total counts are still accurately maintained for the heavy hitters in the frequency list.

Reply:
Created an issue: https://github.com/akka/akka/issues/31093

Reply:
thank, not a priority, I was just curious of how much overlap there was